### PR TITLE
Dimensional Anchor

### PR DESCRIPTION
**Pull request by John-Candlebury**
#### Summary

```SUMMARY: Content "Adds dimensional anchor item and some related fluff content."```

#### Purpose of change
These are quest items for Hub 01 fourth quest. Both the anchor and the suit will be armor to defend against #34168.

We also needed some decent lab finale that better explained the dimensional exploration XEDRA was up to before the end of the world.  

#### Describe the solution
So I added a new lab finale for these items, since it seemed the most logical place to have them. Plus It makes a good target for the fourth hub mission too. The handful of lore documents were added to 
add some flavor, mostly wrote them by the seat of my pants, tell me if they need changing.

#### Describe alternatives you've considered
Not adding the phase immersion suit until there are different dimensions to actually immerse in.

#### Testing
Spawn finale, wear and power anchor, test anchor on the nearby portals.

### CODE REVIEW

File: data/json/items/generic.json [hunk 1]
```diff
@@ -97,6 +97,28 @@
     "volume": "250 ml",
     "flags": [ "TRADER_AVOID" ]
   },
+  {
+    "type": "GENERIC",
+    "id": "log_immersion27",
+    "copy-from": "file",
+    "name": "INCIDENT REPORT: IMMERSION-27A",
+    "description": "A white piece of paper, with the logo of XEDRA printed on its upper left corner. It seems to be an internal report of some kind.",
```
Comment (Night-Pryanik) [inline data/json/items/generic.json]:
```suggestion
    "description": "A white piece of paper, with the logo of XEDRA printed on its upper left corner.  It seems to be an internal report of some kind.",
```

File: data/json/items/generic.json [hunk 2]
```diff
@@ -334,6 +356,21 @@
...
     "color": "white"
   },
+  {
+    "type": "GENERIC",
+    "id": "sample_t_substrate",
+    "category": "spare_parts",
+    "symbol": ",",
+    "color": "light_gray",
+    "name": "t-substrate sample",
+    "description": "An acrylic cube cast around a small black crystal.  Its tepid to the touch.",
```
Comment (Night-Pryanik) [inline data/json/items/generic.json]:
```suggestion
    "description": "An acrylic cube cast around a small black crystal.  It's tepid to the touch.",
```

File: data/json/items/tool_armor.json [hunk 3]
```diff
@@ -645,6 +645,131 @@
...
+  },
+  {
+    "id": "phase_immersion_suit",
+    "type": "TOOL_ARMOR",
+    "category": "clothing",
+    "symbol": "[",
+    "color": "light_gray",
+    "name": "phase immersion suit",
+    "looks_like": "rm13_armor",
+    "description": "Covered in interlocking plates of reflective metal, this advanced suit resembles both plate armor and an astronaut's spacesuit.  Designed to protect its wearer during extra dimensional excursions, it offers unparalleled environmental protection, both in this earth and beyond.  Use it to turn it on",
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
    "description": "Covered in interlocking plates of reflective metal, this advanced suit resembles both plate armor and an astronaut's spacesuit.  Designed to protect its wearer during extra dimensional excursions, it offers unparalleled environmental protection, both in this Earth and beyond.  Use it to turn it on.",
```

File: data/json/items/tool_armor.json [hunk 4]
```diff
@@ -645,6 +645,131 @@
...
+    "material": [ "carbide" ],
+    "weight": "1250 g",
+    "volume": "4500 ml",
+    "storage": "8 L",
+    "ammo": "plutonium",
+    "max_charges": 2500,
+    "initial_charges": 1000,
+    "use_action": {
+      "type": "transform",
+      "msg": "Initiating \nRunning suit integrity diagnostics....\nAll systems operational.",
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
      "msg": "Initiating \nRunning suit integrity diagnostics...\nAll systems operational.",
```

File: data/json/items/tool_armor.json [hunk 5]
```diff
@@ -645,6 +645,131 @@
...
+      "RAINPROOF",
+      "WATCH",
+      "ALARMCLOCK",
+      "SWIM_GOGGLES",
+      "SUN_GLASSES"
+    ]
+  },
+  {
+    "id": "phase_immersion_suit_on",
+    "copy-from": "dimensional_anchor",
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
    "copy-from": "phase_immersion_suit",
```

File: data/json/items/tool_armor.json [hunk 6]
```diff
@@ -645,6 +645,131 @@
...
+    "type": "TOOL_ARMOR",
+    "name": "phase immersion suit (on)",
+    "name_plural": "phase immersion suit (on)",
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
    "name_plural": "phase immersion suits (on)",
```

File: data/json/items/tool_armor.json [hunk 7]
```diff
@@ -645,6 +645,131 @@
...
+    "encumbrance": 2,
+    "material_thickness": 2,
+    "flags": [ "OVERSIZE", "STRAPPED" ]
+  },
+  {
+    "id": "dimensional_anchor_on",
+    "copy-from": "dimensional_anchor",
+    "type": "TOOL_ARMOR",
+    "name": "5-point anchor (on)",
+    "name_plural": "5-point anchor (on)",
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
    "name_plural": "5-point anchors (on)",
```

File: data/json/items/tool_armor.json [hunk 8]
```diff
@@ -645,6 +645,131 @@
...
+      "SUN_GLASSES"
+    ]
+  },
+  {
+    "id": "phase_immersion_suit_on",
+    "copy-from": "dimensional_anchor",
+    "type": "TOOL_ARMOR",
+    "name": "phase immersion suit (on)",
+    "name_plural": "phase immersion suit (on)",
+    "looks_like": "dimensional_anchor",
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
    "looks_like": "phase_immersion_suit",
```

File: data/json/items/tool_armor.json [hunk 9]
```diff
@@ -645,6 +645,131 @@
...
+    "description": "Covered in interlocking plates of reflective metal, this advanced suit resembles both plate armor and an astronaut's spacesuit.  It is turned on, and continually draining power.  Use it to turn it off.",
+    "turns_per_charge": 25,
+    "revert_to": "dimensional_anchor",
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
    "revert_to": "phase_immersion_suit",
```

File: data/json/items/tool_armor.json [hunk 10]
```diff
@@ -645,6 +645,131 @@
...
+    "use_action": { "type": "transform", "menu_text": "Turn off", "msg": "Suit shutting down.", "target": "dimensional_anchor" },
```
Comment (Night-Pryanik) [inline data/json/items/tool_armor.json]:
```suggestion
    "use_action": { "type": "transform", "menu_text": "Turn off", "msg": "Suit shutting down.", "target": "phase_immersion_suit" },
```

File: data/json/snippets/dimensional_exploration.json [hunk 11]
```diff
@@ -0,0 +1,23 @@
+[
+  {
+    "type": "snippet",
+    "category": "log_immersion27",
+    "text": [
+      "\nIMMERSION LOG: 27A\n\nINCIDENT REPORT:\n\nSUMMARY:\nUnforeseen EM interference source forced the team to remain in-phase between prime and sub-scape 27.  Expedition members reported a observing a spheroid-entity ('PE-01') moving beneath the typically opaque t-substrate.  Interference readings and expedition testimony implies the PE-01 was a potent EM radiation source, and the origin of the interference.  After 30min,  PE-01 receded and the expedition was aborted in accordance of procedure EMER-12.  The expedition team suffered no injuries, physical or mental, during the incident. \n\nIts become clear we have greatly underestimated the risk posed by phase immersions.Engineering has be tasked with designing a new, improved protection suite, and has shown promising advances.  Updated emergency procedures are in-progress.\n\n Refer to file EXO-I-271 for full report."
```
Comment (Night-Pryanik) [inline data/json/snippets/dimensional_exploration.json]:
```suggestion
      "\nIMMERSION LOG: 27A\n\nINCIDENT REPORT:\n\nSUMMARY:\nUnforeseen EM interference source forced the team to remain in-phase between prime and sub-scape 27.  Expedition members reported observing a spheroid-entity ('PE-01') moving beneath the typically opaque t-substrate.  Interference readings and expedition testimony implies the PE-01 was a potent EM radiation source, and the origin of the interference.  After 30 min, PE-01 receded and the expedition was aborted in accordance of procedure EMER-12.  The expedition team suffered no injuries, physical or mental, during the incident. \n\nIt's become clear we have greatly underestimated the risk posed by phase immersions.  Engineering has be tasked with designing a new, improved protection suite, and has shown promising advances.  Updated emergency procedures are in-progress.\n\nRefer to file EXO-I-271 for full report."
```

File: data/json/snippets/dimensional_exploration.json [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+    ]
+  },
+  {
+    "type": "snippet",
+    "category": "t-substrate",
+    "text": [
+      "\nANALISIS OF TRANSPLANAR MATERIALS\n\nT-SUBSTRATE\n\nSUMMARY:\nThe material henceforth named t-substrate is the naturally occurring 'floor' material of phase-space. Samples of t-substrate were first recovered during IMMERSION-25A, when upon return to prime the samples violently [////////], exhibiting what are now known as 'plane-uneven' physical properties.  The following report will examine such variances to the best extent of our knowledge.\n\n Refer to file EXO-M-312 for full report."
```
Comment (Night-Pryanik) [inline data/json/snippets/dimensional_exploration.json]:
```suggestion
      "\nANALYSIS OF TRANSPLANAR MATERIALS\n\nT-SUBSTRATE\n\nSUMMARY:\nThe material henceforth named t-substrate is the naturally occurring 'floor' material of phase-space.  Samples of t-substrate were first recovered during IMMERSION-25A, when upon return to prime the samples violently [////////], exhibiting what are now known as 'plane-uneven' physical properties.  The following report will examine such variances to the best extent of our knowledge.\n\nRefer to file EXO-M-312 for full report."
```

File: src/game.cpp [hunk 13]
```diff
@@ -9652,6 +9652,12 @@ bool game::phasing_move( const tripoint &dest_loc )
            ( critter_at( dest ) != nullptr && tunneldist > 0 ) ) {
         //add 1 to tunnel distance for each impassable tile in the line
         tunneldist += 1;
+        //Being dimensionally anchored prevents quantum shenanigans.
+        if( u.worn_with_flag( "DIMENSIONAL_ANCHOR" ) || u.has_effect_with_flag( "DIMENSIONAL_ANCHOR" ) ) {
+            u.add_msg_if_player( m_info, _( "You are repeled by the barrier!" ) );
```
Comment (Night-Pryanik) [inline src/game.cpp]:
```suggestion
            u.add_msg_if_player( m_info, _( "You are repelled by the barrier!" ) );
```

File: src/game.cpp [hunk 14]
```diff
@@ -9652,6 +9652,12 @@ bool game::phasing_move( const tripoint &dest_loc )
...
+            u.charge_power( -250_kJ ); //cost of tunneling one tile.
```
Comment (Night-Pryanik) [inline src/game.cpp]: I guess this should be somehow explained to the player.

File: src/teleport.cpp [hunk 15]
```diff
@@ -18,12 +18,17 @@ bool teleport::teleport( Creature &critter, int min_distance, int max_distance,
         debugmsg( "ERROR: Function teleport::teleport called with invalid arguments." );
         return false;
     }
-
     player *const p = critter.as_player();
     const bool c_is_u = p != nullptr && p == &g->u;
     int tries = 0;
     tripoint origin = critter.pos();
     tripoint new_pos;
+    //The teleportee is dimensionally anchored so nothinng happens
```
Comment (Night-Pryanik) [inline src/teleport.cpp]:
```suggestion
    //The teleportee is dimensionally anchored so nothing happens
```

File: src/teleport.cpp [hunk 16]
```diff
@@ -18,12 +18,17 @@ bool teleport::teleport( Creature &critter, int min_distance, int max_distance,
...
+    if( p && ( p->worn_with_flag( "DIMENSIONAL_ANCHOR" ) ||
+               p->has_effect_with_flag( "DIMENSIONAL_ANCHOR" ) ) ) {
+        p->add_msg_if_player( m_warning, _( "You feel a strange, innwards force." ) );
```
Comment (Night-Pryanik) [inline src/teleport.cpp]:
```suggestion
        p->add_msg_if_player( m_warning, _( "You feel a strange, inwards force." ) );
```

File: src/teleport.cpp [hunk 17]
```diff
@@ -49,16 +54,22 @@ bool teleport::teleport( Creature &critter, int min_distance, int max_distance,
...
     if( Creature *const poor_soul = g->critter_at<Creature>( new_pos ) ) {
+        player *const poor_player = dynamic_cast<player *>( poor_soul );
         if( safe ) {
             if( c_is_u ) {
                 add_msg( m_bad, _( "You cannot teleport safely." ) );
             }
             return false;
+        } else if( poor_player && ( poor_player->worn_with_flag( "DIMENSIONAL_ANCHOR" ) ||
+                                    poor_player->has_effect_with_flag( "DIMENSIONAL_ANCHOR" ) ) ) {
+            debugmsg( "ERROR: 3." );
```
Comment (Night-Pryanik) [inline src/teleport.cpp]: ?

File: src/game.cpp (hunk 14 above)
Comment (KorGgenT) [inline src/game.cpp]: this function was migrated recently. 
```suggestion
            u.mod_power_level( -250_kJ ); //cost of tunneling one tile.
```
you may need to rebase on current master.

**Comment by simmael** [comment]
Just a little typo a suppose, immersion suit after activating change into anchor. There is wrong target in json, instead of immersion_suit_on, its probably coppied and forgotten from dimensional_anchor

**Comment by John-Candlebury** [comment]
Thanks for letting me know

**Comment by simmael** [comment]
no problem. it should be me saying thanks for making content to the game.

File: data/json/items/tool_armor.json [hunk 18]
```diff
@@ -645,6 +645,130 @@
...
+    "initial_charges": 1000,
+    "use_action": {
+      "type": "transform",
+      "msg": "Initiating \nRunning suit integrity diagnostics...\nAll systems operational.",
+      "target": "dimensional_anchor_on",
+      "active": true,
+      "need_charges": 1,
+      "need_charges_msg": "Warning: Operating on minimal power.  Protection compromised."
+    },
+    "covers": [ "TORSO" ],
```
Comment (chaisawlajatang) [inline data/json/items/tool_armor.json]: Looks like it should cover the entire body, isn't it?

**Comment by RDru** [comment]
I randomly stumbled upon this lab part. I wish you would add more lore from #34168 into the game.

**Pull request merged**